dwave_networkx/algorithms/partition.py

```diff
@@ -0,0 +1,157 @@
+    >>> G = dnx.chimera_graph(1, 1, 4)
+    >>> partitions = dnx.partition(G, sampler=sampler)
+    Notes
+    -----
+    Samplers by their nature may not return the optimal solution. This
+    function does not attempt to confirm the quality of the returned
+    sample.
+    """
+
+    cqm = graph_partition_cqm(G, num_partitions)
+    
+    if not len(G.nodes):
+        return {}
+
+    # Solve the problem using the CQM solver
+    response = sampler.sample_cqm(cqm, **sampler_args)
+
+    # Consider only results satisfying all constraints
+    possible_partitions = response.filter(lambda d: d.is_feasible)
+    
+    if not possible_partitions: 
+        raise Exception("No feasible solution could be found for this problem instance.")
+
+    # Reinterpret result as partition assignment over nodes
+    indicators = (key for key, value in possible_partitions.first.sample.items() if isclose(value, 1.))
+    node_partition = {key[0]: key[1] for key in indicators}
+    
+    return node_partition
+    
+    
+def graph_partition_cqm(G, num_partitions):
+    """Find a constrained quadratic model for the graph's partitions.
+
+    Defines an CQM with ground states corresponding to a
+    balanced k-partition of G and uses the sampler to sample from it.
+    A k-partition is a collection of k subsets of the vertices
+    of G such that each vertex is in exactly one subset, and
+    the number of edges between vertices in different subsets
+    is as small as possible. If G is a weighted graph, the sum
+    of weights over those edges are minimized.
+    Parameters
+    ----------
+    G : NetworkX graph
+        The graph to partition.
+    num_partitions : int
+        The number of subsets in the desired partition.
+
+
+    Returns
+    -------
+    cqm : :class:`dimod.ConstrainedQuadraticModel`
+        A constrained quadratic model with ground states corresponding to a
+        partition problem. The nodes of `G` are discrete logical variables
+        of the CQM, where the cases are the different partitions the node
+        can be assigned to. The objective is given as the number of edges
+        connecting nodes in different partitions.
+
+    """
+    partition_size = G.number_of_nodes()/num_partitions
+    partitions = range(num_partitions)
+    cqm = ConstrainedQuadraticModel()
+
+    # Variables will be added using the discrete method in CQM    
+    x = {(v,k): Binary((v,k)) for v,k in product(G.nodes, partitions)}
+    
+    for v in G.nodes:
+        cqm.add_discrete([(v, k) for k in partitions], label=v)
+        
+        
+    if not isclose(partition_size, int(partition_size)):
+        # if number of nodes don't divide into num_partitions,
+        # accept partitions of size ceil() or floor()
+        floor, ceil = int(partition_size), int(partition_size+1)
+        for k in partitions:
+            cqm.add_constraint(quicksum([x[u, k] for u in G.nodes]) >= floor, label='equal_partition_low_%s' %k)

```

Comment:
Can use a generator rather than a list. I.e. `(x[u, k] for u in G.nodes)` rather than `[x[u, k] for u in G.nodes]`